issue with mypy

### Description
<!-- Give a clear and concise description of the bug: -->
On current master, `mypy` throws errors for me but not for @giovp 
I have:
- cloned squidpy
- installed pre-commit `pre-commit clean; pre-commit install`
- changed docstring of a file
- done `git commit` which results in the error

Is there an issue with my environment?

### Minimal reproducible example
<!-- Put an example code with which we can reproduce the bug: -->
```python
...
```

### Traceback
<!-- If applicable, replace `...` with an error traceback below: -->

<details>

```pytb
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

squidpy/gr/_utils.py:86:16: error: Item "ndarray" of "Union[ndarray, Any]" has no attribute "tocsc"  [union-attr]
squidpy/gr/_ppatterns.py:284: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ppatterns.py:285: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ppatterns.py:286: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ppatterns.py:287: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ppatterns.py:288: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:662: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:663: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:708:17: error: Incompatible types in assignment (expression has type "ndarray", variable has type "List[ndarray]")  [assignment]
squidpy/gr/_ligrec.py:710:31: error: Subclass of "List[ndarray]" and "ndarray" cannot exist: would have incompatible method signatures  [unreachable]
squidpy/gr/_ligrec.py:712:9: error: Statement is unreachable  [unreachable]
squidpy/gr/_ligrec.py:746: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:747: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:748: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:749: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:750: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:751: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:752: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:817:21: error: Incompatible types in assignment (expression has type "None", variable has type "ndarray")  [assignment]
squidpy/gr/_build.py:186:5: error: Returning Any from function declared to return "ndarray"  [no-any-return]
squidpy/gr/_build.py:186:12: error: Item "ndarray" of "Union[Any, ndarray]" has no attribute "multiply"  [union-attr]
squidpy/pl/_utils.py:205:15: error: Item "ndarray" of "Union[Any, ndarray]" has no attribute "A"  [union-attr]
squidpy/pl/_utils.py:212:5: error: Returning Any from function declared to return "ndarray"  [no-any-return]
squidpy/pl/_utils.py:239:19: error: Item "ndarray" of "Union[Any, ndarray]" has no attribute "toarray"  [union-attr]
squidpy/pl/_utils.py:562: error: Function "numpy.array" is not valid as a type  [valid-type]
squidpy/pl/_utils.py:562: note: Perhaps you need "Callable[...]" or a callback protocol?
squidpy/pl/_utils.py:572:28: error: np.array? has no attribute "T"  [attr-defined]
squidpy/im/_utils.py:17:5: error: Returning Any from function declared to return "ndarray"  [no-any-return]
squidpy/im/_feature_mixin.py:342:13: error: Returning Any from function declared to return "ndarray"  [no-any-return]
squidpy/im/_feature_mixin.py:370:56: error: Incompatible types in assignment (expression has type "ndarray", target has type "int")  [assignment]
squidpy/im/_feature_mixin.py:374:56: error: Incompatible types in assignment (expression has type "number[Any]", target has type "int")  [assignment]
squidpy/im/_feature_mixin.py:375:55: error: Incompatible types in assignment (expression has type "number[Any]", target has type "int")  [assignment]
squidpy/im/_feature_mixin.py:387:63: error: Incompatible types in assignment (expression has type "number[Any]", target has type "int")  [assignment]
squidpy/im/_feature_mixin.py:388:62: error: Incompatible types in assignment (expression has type "number[Any]", target has type "int")  [assignment]
squidpy/im/_container.py:833:19: error: Incompatible types in assignment (expression has type "ImageContainer", variable has type "ndarray")  [assignment]
squidpy/im/_container.py:834:13: error: "memoryview" has no attribute "attrs"  [attr-defined]
squidpy/im/_container.py:836:20: error: Incompatible return value type (got "ndarray", expected "Optional[ImageContainer]")  [return-value]
squidpy/im/_container.py:836: error: unused 'type: ignore' comment
squidpy/im/_segment.py:146:9: error: Returning Any from function declared to return "ndarray"  [no-any-return]
squidpy/im/_segment.py:166:9: error: Returning Any from function declared to return "ndarray"  [no-any-return]
squidpy/pl/_interactive/_controller.py:192:9: error: Returning Any from function declared to return "ndarray"  [no-any-return]
Found 40 errors in 10 files (checked 1 source file)
```

</details>

### Version
<!-- Output of squidpy.__version__ -->
...

I have mypy version `0.790` (`mypy --version`) and have tested `0.8`, running `pre-commit run --all-files` works.
I am using Python 3.8.5. The errors of type:
```bash
squidpy/gr/_ppatterns.py:284: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
```
come from `data-science-types`, but we use `from __future__ import annotations` at the top, so there should be no problem.
I'd try a creatinga fresh new env.

> I have mypy version `0.790` (`mypy --version`) and have tested `0.8`, running `pre-commit run --all-files` works.
> I am using Python 3.8.5. The errors of type:
> 
> ```shell
> squidpy/gr/_ppatterns.py:284: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
> ```
> 
> come from `data-science-types`, but we use `from __future__ import annotations` at the top, so there should be no problem.
> I'd try a creatinga fresh new env.

With fresh env you mean a new python virtualenv / conda where I install squidpy in? I've already deleted `.tox` several times and re-installed squidpy. Will try with a new environment.

Hm, still no luck. With a completely new Conda (with python 3.8), and only squidpy installed (`pip install -e'.[dev,test]'`), I get the same error

<details>
> tox -e lint
lint create: /Users/hannah.spitzer/projects/spatial_scanpy/squidpy/.tox/lint
lint installdeps: pre-commit>=2.9.3
lint installed: appdirs==1.4.4,cfgv==3.2.0,distlib==0.3.1,filelock==3.0.12,identify==1.5.13,nodeenv==1.5.0,pre-commit==2.10.1,PyYAML==5.4.1,six==1.15.0,toml==0.10.2,virtualenv==20.4.2
lint run-test-pre: PYTHONHASHSEED='610168850'
lint run-test: commands[0] | pre-commit run --all-files --show-diff-on-failure
[INFO] Initializing environment for https://github.com/pre-commit/mirrors-mypy.
[INFO] Initializing environment for https://github.com/pre-commit/mirrors-mypy:numpy>=1.15.0,pandas,scipy>=1.5.0.
[INFO] Initializing environment for https://github.com/psf/black.
[INFO] Initializing environment for https://github.com/psf/black:toml.
[INFO] Initializing environment for https://github.com/timothycrosley/isort.
[INFO] Initializing environment for https://github.com/timothycrosley/isort:toml.
[INFO] Initializing environment for https://github.com/asottile/yesqa.
[INFO] Initializing environment for https://github.com/asottile/yesqa:flake8-blind-except,flake8-bugbear,flake8-builtins,flake8-comprehensions,flake8-docstrings,flake8-logging-format,flake8-mock,flake8-pytest-style,flake8-rst-docstrings,flake8-string-format,flake8-tidy-imports.
[INFO] Initializing environment for https://github.com/macisamuele/language-formatters-pre-commit-hooks.
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/myint/autoflake.
[INFO] Initializing environment for https://gitlab.com/pycqa/flake8.
[INFO] Initializing environment for https://gitlab.com/pycqa/flake8:flake8-blind-except,flake8-bugbear,flake8-builtins,flake8-comprehensions,flake8-docstrings,flake8-logging-format,flake8-mock,flake8-pytest-style,flake8-rst-docstrings,flake8-string-format,flake8-tidy-imports.
[INFO] Initializing environment for https://github.com/jumanjihouse/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/myint/rstcheck.
[INFO] Initializing environment for https://github.com/asottile/blacken-docs.
[INFO] Initializing environment for https://github.com/asottile/blacken-docs:black==20.8b1.
[INFO] Initializing environment for https://github.com/asottile/pyupgrade.
[INFO] Initializing environment for https://github.com/pre-commit/pygrep-hooks.
[INFO] Initializing environment for https://github.com/PyCQA/doc8.
[INFO] Installing environment for https://github.com/pre-commit/mirrors-mypy.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/psf/black.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/timothycrosley/isort.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/asottile/yesqa.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/macisamuele/language-formatters-pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/myint/autoflake.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://gitlab.com/pycqa/flake8.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/myint/rstcheck.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/asottile/blacken-docs.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/asottile/pyupgrade.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/PyCQA/doc8.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

squidpy/gr/_utils.py:86:16: error: Item "ndarray" of "Union[ndarray, Any]" has no attribute "tocsc"  [union-attr]
squidpy/pl/_utils.py:205:15: error: Item "ndarray" of "Union[Any, ndarray]" has no attribute "A"  [union-attr]
squidpy/pl/_utils.py:212:5: error: Returning Any from function declared to return "ndarray"  [no-any-return]
squidpy/pl/_utils.py:239:19: error: Item "ndarray" of "Union[Any, ndarray]" has no attribute "toarray"  [union-attr]
squidpy/pl/_utils.py:562: error: Function "numpy.array" is not valid as a type  [valid-type]
squidpy/pl/_utils.py:562: note: Perhaps you need "Callable[...]" or a callback protocol?
squidpy/pl/_utils.py:572:28: error: np.array? has no attribute "T"  [attr-defined]
squidpy/gr/_ppatterns.py:284: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ppatterns.py:285: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ppatterns.py:286: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ppatterns.py:287: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ppatterns.py:288: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_build.py:186:5: error: Returning Any from function declared to return "ndarray"  [no-any-return]
squidpy/gr/_build.py:186:12: error: Item "ndarray" of "Union[Any, ndarray]" has no attribute "multiply"  [union-attr]
squidpy/gr/_ligrec.py:662: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:663: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:708:17: error: Incompatible types in assignment (expression has type "ndarray", variable has type "List[ndarray]")  [assignment]
squidpy/gr/_ligrec.py:710:31: error: Subclass of "List[ndarray]" and "ndarray" cannot exist: would have incompatible method signatures  [unreachable]
squidpy/gr/_ligrec.py:712:9: error: Statement is unreachable  [unreachable]
squidpy/gr/_ligrec.py:746: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:747: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:748: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:749: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:750: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:751: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:752: error: "ndarray" expects no type arguments, but 1 given  [type-arg]
squidpy/gr/_ligrec.py:817:21: error: Incompatible types in assignment (expression has type "None", variable has type "ndarray")  [assignment]
squidpy/im/_utils.py:17:5: error: Returning Any from function declared to return "ndarray"  [no-any-return]
squidpy/im/_feature_mixin.py:342:13: error: Returning Any from function declared to return "ndarray"  [no-any-return]
squidpy/im/_feature_mixin.py:370:56: error: Incompatible types in assignment (expression has type "ndarray", target has type "int")  [assignment]
squidpy/im/_feature_mixin.py:374:56: error: Incompatible types in assignment (expression has type "number[Any]", target has type "int")  [assignment]
squidpy/im/_feature_mixin.py:375:55: error: Incompatible types in assignment (expression has type "number[Any]", target has type "int")  [assignment]
squidpy/im/_feature_mixin.py:387:63: error: Incompatible types in assignment (expression has type "number[Any]", target has type "int")  [assignment]
squidpy/im/_feature_mixin.py:388:62: error: Incompatible types in assignment (expression has type "number[Any]", target has type "int")  [assignment]
squidpy/im/_container.py:833:19: error: Incompatible types in assignment (expression has type "ImageContainer", variable has type "ndarray")  [assignment]
squidpy/im/_container.py:834:13: error: "memoryview" has no attribute "attrs"  [attr-defined]
squidpy/im/_container.py:836: error: unused 'type: ignore' comment
squidpy/im/_container.py:836:20: error: Incompatible return value type (got "ndarray", expected "Optional[ImageContainer]")  [return-value]
squidpy/im/_segment.py:146:9: error: Returning Any from function declared to return "ndarray"  [no-any-return]
squidpy/im/_segment.py:166:9: error: Returning Any from function declared to return "ndarray"  [no-any-return]
squidpy/pl/_interactive/_controller.py:192:9: error: Returning Any from function declared to return "ndarray"  [no-any-return]
Found 40 errors in 10 files (checked 58 source files)

black....................................................................Passed
isort....................................................................Passed
Strip unnecessary `# noqa`s..............................................Passed
Pretty format YAML.......................................................Passed
Detect Private Key.......................................................Passed
Check for merge conflicts................................................Passed
Check python ast.........................................................Passed
Check for broken symlinks............................(no files to check)Skipped
Check for added large files..............................................Passed
Check that executables have shebangs.....................................Passed
Fix python encoding pragma...............................................Passed
Fix End of Files.........................................................Passed
Mixed line ending........................................................Passed
Trim Trailing Whitespace.................................................Passed
Tests should end in _test.py.............................................Passed
Check for case conflicts.................................................Passed
Check docstring is first.................................................Passed
Check Yaml...............................................................Passed
Check Toml...............................................................Passed
Fix requirements.txt.....................................................Passed
autoflake................................................................Passed
flake8...................................................................Passed
Check executable files use .sh extension.................................Passed
rstcheck.................................................................Passed
blacken-docs.............................................................Passed
pyupgrade................................................................Passed
check for eval().........................................................Passed
type annotations not comments............................................Passed
check blanket noqa.......................................................Passed
rst ``code`` is two backticks............................................Passed
rst directives end with two colons.......................................Passed
rst ``inline code`` next to normal text..................................Passed
doc8.....................................................................Passed
ERROR: InvocationError for command /Users/hannah.spitzer/projects/spatial_scanpy/squidpy/.tox/lint/bin/pre-commit run --all-files --show-diff-on-failure (exited with code 1)
</details>

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -60,7 +60,7 @@ jobs:
     test:
         needs: init
         if: ${{ github.event_name == 'schedule' || needs.init.outputs.skip == 'false' }}
-        timeout-minutes: 10
+        timeout-minutes: 20
         runs-on: ${{ matrix.os }}
         strategy:
             fail-fast: false

--- a/.mypy.ini
+++ b/.mypy.ini
@@ -1,6 +1,6 @@
 [mypy]
 mypy_path = squidpy
-python_version = 3.7
+python_version = 3.8
 
 warn_redundant_casts = True
 warn_unused_configs = True

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -10,7 +10,7 @@ repos:
     rev: v0.800
     hooks:
     -   id: mypy
-        additional_dependencies: [numpy>=1.15.0, scipy>=1.5.0, pandas]
+        additional_dependencies: [numpy, scipy, pandas]
 -   repo: https://github.com/psf/black
     rev: 20.8b1
     hooks:

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,6 +1,5 @@
 anndata>=0.7.4
 dask>=2.30
-data-science-types>=0.2.20  # array typehints
 docrep>=0.3.1
 ipywidgets>=7.5.1  # progress bar
 omnipath>=1.0.4

--- a/squidpy/_utils.py
+++ b/squidpy/_utils.py
@@ -28,11 +28,11 @@ __all__ = ["singledispatchmethod", "Signal", "SigQueue"]
 
 
 try:
-    from functools import singledispatchmethod  # type: ignore[attr-defined]
+    from functools import singledispatchmethod
 except ImportError:
     from functools import singledispatch, update_wrapper
 
-    def singledispatchmethod(func: Callable[..., Any]) -> Callable[..., Any]:
+    def singledispatchmethod(func: Callable[..., Any]) -> Callable[..., Any]:  # type: ignore[no-redef]
         """Backport of `singledispatchmethod` for < Python 3.8."""
         dispatcher = singledispatch(func)
 

--- a/squidpy/gr/_build.py
+++ b/squidpy/gr/_build.py
@@ -5,7 +5,8 @@ import warnings
 from scanpy import logging as logg
 from anndata import AnnData
 
-from scipy.sparse import spmatrix, csr_matrix, SparseEfficiencyWarning
+from numba import njit
+from scipy.sparse import spmatrix, csr_matrix, isspmatrix_csr, SparseEfficiencyWarning
 from sklearn.neighbors import NearestNeighbors
 from sklearn.metrics.pairwise import cosine_similarity
 import numpy as np
@@ -114,11 +115,12 @@ def spatial_neighbors(
     if transform == Transform.SPECTRAL:
         Adj = _transform_a_spectral(Adj)
     elif transform == Transform.COSINE:
+        print("foo")
         Adj = _transform_a_cosine(Adj)
     elif transform == Transform.NONE:
         pass
     else:
-        raise NotImplementedError(transform)
+        raise NotImplementedError(f"Transform `{transform}` is not yet implemented.")
 
     neighs_key = Key.uns.spatial_neighs(key_added)
     conns_key = Key.obsp.spatial_conn(key_added)
@@ -181,10 +183,28 @@ def _build_connectivity(
     return (conns_m, dists_m) if return_distance else conns_m
 
 
-def _transform_a_spectral(a: Union[csr_matrix, np.ndarray]) -> np.ndarray:
+@njit
+def outer(indices: np.ndarray, indptr: np.ndarray, degrees: np.ndarray) -> np.ndarray:
+    res = np.empty_like(indices, dtype=np.float64)
+    start = 0
+    for i in range(len(indptr) - 1):
+        ixs = indices[indptr[i] : indptr[i + 1]]
+        res[start : start + len(ixs)] = degrees[i] * degrees[ixs]
+        start += len(ixs)
+
+    return res
+
+
+def _transform_a_spectral(a: spmatrix) -> spmatrix:
+    if not isspmatrix_csr(a):
+        a = a.tocsr()
     degrees = np.squeeze(np.array(np.sqrt(1.0 / a.sum(axis=0))))
-    return a.multiply(np.outer(degrees, degrees))
+
+    a = a.multiply(outer(a.indices, a.indptr, degrees))
+    a.eliminate_zeros()
+
+    return a
 
 
-def _transform_a_cosine(a: Union[spmatrix, np.ndarray]) -> Union[np.ndarray, spmatrix]:
+def _transform_a_cosine(a: spmatrix) -> spmatrix:
     return cosine_similarity(a, dense_output=False)

--- a/squidpy/gr/_ligrec.py
+++ b/squidpy/gr/_ligrec.py
@@ -703,13 +703,13 @@ def _analysis(
     def extractor(res: Sequence[TempResult]) -> TempResult:
         assert len(res) == n_jobs, f"Expected to find `{n_jobs}` results, found `{len(res)}`."
 
-        means: List[np.ndarray] = [r.means for r in res if r.means is not None]
-        assert len(means) == 1, f"Only `1` job should've calculated the means, but found `{len(means)}`."
-        means = means[0]
+        meanss: List[np.ndarray] = [r.means for r in res if r.means is not None]
+        assert len(meanss) == 1, f"Only `1` job should've calculated the means, but found `{len(meanss)}`."
+        means = meanss[0]
         if TYPE_CHECKING:
             assert isinstance(means, np.ndarray)
 
-        pvalues: np.ndarray = np.sum([r.pvalues for r in res if r.pvalues is not None], axis=0) / float(n_perms)
+        pvalues = np.sum([r.pvalues for r in res if r.pvalues is not None], axis=0) / float(n_perms)
         assert means.shape == pvalues.shape, f"Means and p-values differ in shape: `{means.shape}`, `{pvalues.shape}`."
 
         return TempResult(means=means, pvalues=pvalues)
@@ -811,7 +811,7 @@ def _analysis_helper(
     _test = globals()[fn_key]
 
     if return_means:
-        res_means = np.zeros((len(interactions), len(interaction_clusters)), dtype=np.float64)
+        res_means: Optional[np.ndarray] = np.zeros((len(interactions), len(interaction_clusters)), dtype=np.float64)
         test = partial(_test, res_means=res_means)
     else:
         res_means = None

--- a/squidpy/gr/_ligrec.py
+++ b/squidpy/gr/_ligrec.py
@@ -659,8 +659,8 @@ def ligrec(
 @d.dedent
 def _analysis(
     data: pd.DataFrame,
-    interactions: np.ndarray[np.uint32],
-    interaction_clusters: np.ndarray[np.uint32],
+    interactions: np.ndarray,
+    interaction_clusters: np.ndarray,
     threshold: float = 0.1,
     n_perms: int = 1000,
     seed: Optional[int] = None,
@@ -743,13 +743,13 @@ def _analysis(
 
 
 def _analysis_helper(
-    perms: np.ndarray[np.uint32],
-    data: np.ndarray[np.float64],
-    mean: np.ndarray[np.float64],
-    mask: np.ndarray[np.bool_],
-    interactions: np.ndarray[np.uint32],
-    interaction_clusters: np.ndarray[np.uint32],
-    clustering: np.ndarray[np.uint32],
+    perms: np.ndarray,
+    data: np.ndarray,
+    mean: np.ndarray,
+    mask: np.ndarray,
+    interactions: np.ndarray,
+    interaction_clusters: np.ndarray,
+    clustering: np.ndarray,
     seed: Optional[int] = None,
     numba_parallel: Optional[bool] = None,
     queue: Optional[SigQueue] = None,

--- a/squidpy/gr/_ppatterns.py
+++ b/squidpy/gr/_ppatterns.py
@@ -1,6 +1,4 @@
 """Functions for point patterns spatial statistics."""
-from __future__ import annotations
-
 from typing import Tuple, Union, Iterable, Optional, Sequence
 from itertools import chain
 from typing_extensions import Literal  # < 3.8
@@ -275,11 +273,11 @@ def _set_weight_class(adata: AnnData, key: str) -> W:
     fastmath=True,
 )
 def _occur_count(
-    clust: Tuple[np.ndarray[np.int32], np.ndarray[np.int32]],
-    pw_dist: np.ndarray[np.float32],
-    labs_unique: np.ndarray[np.int32],
-    interval: np.ndarray[np.float32],
-) -> np.ndarray[np.float32]:
+    clust: Tuple[np.ndarray, np.ndarray],
+    pw_dist: np.ndarray,
+    labs_unique: np.ndarray,
+    interval: np.ndarray,
+) -> np.ndarray:
     num = labs_unique.shape[0]
     out = np.zeros((num, num, interval.shape[0] - 1), dtype=ft)
 

--- a/squidpy/gr/_ppatterns.py
+++ b/squidpy/gr/_ppatterns.py
@@ -1,23 +1,9 @@
 """Functions for point patterns spatial statistics."""
 from __future__ import annotations
 
-from squidpy.gr._utils import (
-    _save_data,
-    _assert_positive,
-    _assert_spatial_basis,
-    _assert_categorical_obs,
-    _assert_connectivity_key,
-    _assert_non_empty_sequence,
-)
-
-try:
-    # [ py<3.8 ]
-    from typing import Literal  # type: ignore[attr-defined]
-except ImportError:
-    from typing_extensions import Literal
-
 from typing import Tuple, Union, Iterable, Optional, Sequence
 from itertools import chain
+from typing_extensions import Literal  # < 3.8
 import warnings
 
 from scanpy import logging as logg
@@ -33,6 +19,14 @@ import numba.types as nt
 
 from squidpy._docs import d, inject_docs
 from squidpy._utils import Signal, SigQueue, parallelize, _get_n_cores
+from squidpy.gr._utils import (
+    _save_data,
+    _assert_positive,
+    _assert_spatial_basis,
+    _assert_categorical_obs,
+    _assert_connectivity_key,
+    _assert_non_empty_sequence,
+)
 from squidpy._constants._pkg_constants import Key
 
 try:
@@ -145,7 +139,7 @@ def moran(
     adata: AnnData,
     connectivity_key: str = Key.obsp.spatial_conn(),
     genes: Optional[Union[str, Sequence[str]]] = None,
-    transformation: Literal["r", "B", "D", "U", "V"] = "r",  # type: ignore[name-defined]
+    transformation: Literal["r", "B", "D", "U", "V"] = "r",
     n_perms: int = 1000,
     corr_method: Optional[str] = "fdr_bh",
     layer: Optional[str] = None,
@@ -236,7 +230,7 @@ def _moran_helper(
     gen: Iterable[str],
     adata: AnnData,
     weights: W,
-    transformation: Literal["r", "B", "D", "U", "V"] = "B",  # type: ignore[name-defined]
+    transformation: Literal["r", "B", "D", "U", "V"] = "r",
     permutations: int = 1000,
     layer: Optional[str] = None,
     seed: Optional[int] = None,

--- a/squidpy/gr/_utils.py
+++ b/squidpy/gr/_utils.py
@@ -81,7 +81,6 @@ def _create_sparse_df(
     -------
     Each column of the DataFrame is stored as a :class:`arrays.SparseArray`.
     """
-    from pandas import DataFrame
     from pandas._libs.sparse import IntIndex
     from pandas.core.arrays.sparse.accessor import (
         SparseArray,
@@ -120,7 +119,7 @@ def _create_sparse_df(
         arr = SparseArray._simple_new(array_data[sl], idx, dtype)
         arrays.append(arr)
 
-    return DataFrame._from_arrays(arrays, columns=columns, index=index, verify_integrity=False)
+    return pd.DataFrame._from_arrays(arrays, columns=columns, index=index, verify_integrity=False)
 
 
 def _assert_categorical_obs(adata: AnnData, key: str) -> None:

--- a/squidpy/gr/_utils.py
+++ b/squidpy/gr/_utils.py
@@ -1,5 +1,15 @@
 """Graph utilities."""
-from typing import Any, List, Tuple, Union, Hashable, Iterable, Optional, Sequence
+from typing import (
+    Any,
+    List,
+    Tuple,
+    Union,
+    Hashable,
+    Iterable,
+    Optional,
+    Sequence,
+    TYPE_CHECKING,
+)
 
 from scanpy import logging as logg
 from anndata import AnnData
@@ -83,6 +93,8 @@ def _create_sparse_df(
         data = csc_matrix(data)
         sort_indices = False
     else:
+        if TYPE_CHECKING:
+            assert isinstance(data, spmatrix)
         data = data.tocsc()
         sort_indices = True
 

--- a/squidpy/im/_container.py
+++ b/squidpy/im/_container.py
@@ -16,15 +16,25 @@ from typing import (
     Sequence,
     TYPE_CHECKING,
 )
+from pathlib import Path
 from itertools import chain
+from typing_extensions import Literal
 import re
+
+from scanpy import logging as logg
+from anndata import AnnData
+
+import numpy as np
+import xarray as xr
 
 import matplotlib as mpl
 import matplotlib.pyplot as plt
 
+from imageio import imread
 from skimage.util import img_as_float
 from skimage.transform import rescale
 
+from squidpy._docs import d
 from squidpy._utils import singledispatchmethod
 from squidpy.gr._utils import (
     _assert_in_range,
@@ -33,23 +43,6 @@ from squidpy.gr._utils import (
     _assert_spatial_basis,
     _assert_non_empty_sequence,
 )
-
-try:
-    from typing import Literal  # type: ignore[attr-defined]
-except ImportError:
-    from typing_extensions import Literal
-
-from pathlib import Path
-
-from scanpy import logging as logg
-from anndata import AnnData
-
-import numpy as np
-import xarray as xr
-
-from imageio import imread
-
-from squidpy._docs import d
 from squidpy.im._utils import (
     _num_pages,
     CropCoords,
@@ -830,10 +823,10 @@ class ImageContainer(FeatureMixin):
             res = res[..., np.newaxis]
 
         if copy:
-            res = ImageContainer(res, layer=layer, channel_dim=channel_dim)
-            res.data.attrs = self.data.attrs.copy()
+            cont = ImageContainer(res, layer=layer, channel_dim=channel_dim)
+            cont.data.attrs = self.data.attrs.copy()
 
-            return res  # type: ignore[no-any-return]
+            return cont
 
         self.add_img(
             res,

--- a/squidpy/im/_feature_mixin.py
+++ b/squidpy/im/_feature_mixin.py
@@ -337,7 +337,7 @@ class FeatureMixin:
             y = coord.slice[0].start + (y_slc.stop - y_slc.start) * y
             x = coord.slice[1].start + (x_slc.stop - x_slc.start) * x
 
-            return np.c_[x, y]
+            return np.c_[x, y]  # type: ignore[no-any-return]
 
         label_layer = self._get_layer(label_layer)
 
@@ -356,7 +356,7 @@ class FeatureMixin:
         else:
             channels = ()
 
-        features = {}
+        features: Dict[str, Any] = {}
         # calculate features that do not depend on the intensity image
         tmp_features = skimage.measure.regionprops_table(
             self[label_layer].values[:, :, 0], properties=no_intensity_props

--- a/squidpy/im/_segment.py
+++ b/squidpy/im/_segment.py
@@ -143,7 +143,7 @@ class SegmentationWatershed(SegmentationModel):
 
         markers, _ = ndi.label(local_maxi)
 
-        return watershed(arr, markers, mask=mask)
+        return np.asarray(watershed(arr, markers, mask=mask))
 
 
 class SegmentationCustom(SegmentationModel):
@@ -163,7 +163,7 @@ class SegmentationCustom(SegmentationModel):
         super().__init__(model=func)
 
     def _segment(self, arr: np.ndarray, **kwargs: Any) -> np.ndarray:
-        return self._model(arr, **kwargs)
+        return np.asarray(self._model(arr, **kwargs))
 
     def __repr__(self) -> str:
         return f"{self.__class__.__name__}[function={getattr(self._model, '__name__', None)}]"

--- a/squidpy/im/_utils.py
+++ b/squidpy/im/_utils.py
@@ -13,8 +13,7 @@ from squidpy.gr._utils import _assert_non_negative
 
 def _circular_mask(arr: np.ndarray, y: int, x: int, radius: float) -> np.ndarray:
     Y, X = np.ogrid[: arr.shape[0], : arr.shape[1]]
-
-    return ((Y - y) ** 2 + (X - x) ** 2) <= radius ** 2
+    return np.asarray(((Y - y) ** 2 + (X - x) ** 2) <= radius ** 2)
 
 
 def _num_pages(fname: str) -> int:

--- a/squidpy/pl/_graph.py
+++ b/squidpy/pl/_graph.py
@@ -2,17 +2,11 @@
 
 from types import MappingProxyType
 from typing import Any, Tuple, Union, Mapping, Optional, Sequence
-
-from scanpy.plotting._utils import add_colors_for_categorical_sample_annotation
-
-try:
-    from typing import Literal  # type: ignore[attr-defined]
-except ImportError:
-    from typing_extensions import Literal
-
 from pathlib import Path
+from typing_extensions import Literal
 
 from anndata import AnnData
+from scanpy.plotting._utils import add_colors_for_categorical_sample_annotation
 
 import numpy as np
 import pandas as pd
@@ -202,7 +196,7 @@ def interaction_matrix(
 def nhood_enrichment(
     adata: AnnData,
     cluster_key: str,
-    mode: Literal["zscore", "count"] = "zscore",  # type: ignore[name-defined]
+    mode: Literal["zscore", "count"] = "zscore",
     annotate: bool = False,
     method: Optional[str] = None,
     title: Optional[str] = None,

--- a/squidpy/pl/_interactive/_controller.py
+++ b/squidpy/pl/_interactive/_controller.py
@@ -189,7 +189,7 @@ class ImageController:
         -------
         The screenshot as an RGB array of shape `(height, width, 3)`.
         """
-        return np.asanyarray(self.view.viewer.screenshot(path, canvas_only=True))
+        return np.asarray(self.view.viewer.screenshot(path, canvas_only=True))
 
     def _handle_already_present(self, layer_name: str) -> None:
         logg.warning(f"Layer `{layer_name}` is already loaded")

--- a/squidpy/pl/_interactive/_controller.py
+++ b/squidpy/pl/_interactive/_controller.py
@@ -187,7 +187,7 @@ class ImageController:
 
         Returns
         -------
-        The screenshot as an RGB array of shape `(height, width, 3)`.
+        Screenshot as an RGB array of shape ``(height, width, 3)``.
         """
         return np.asarray(self.view.viewer.screenshot(path, canvas_only=True))
 

--- a/squidpy/pl/_interactive/_model.py
+++ b/squidpy/pl/_interactive/_model.py
@@ -15,7 +15,7 @@ __all__ = ["ImageModel"]
 
 @dataclass
 class ImageModel:
-    """Model which holds the data for _interactive visualization."""
+    """Model which holds the data for interactive visualization."""
 
     adata: AnnData
     container: ImageContainer

--- a/squidpy/pl/_interactive/_utils.py
+++ b/squidpy/pl/_interactive/_utils.py
@@ -12,6 +12,8 @@ import pandas as pd
 
 from matplotlib.colors import to_rgb
 
+from squidpy._constants._pkg_constants import Key
+
 
 def _get_categorical(
     adata: AnnData,
@@ -30,7 +32,7 @@ def _get_categorical(
     add_colors_for_categorical_sample_annotation(
         adata, key=key, force_update_colors=palette is not None, palette=palette
     )
-    col_dict = dict(zip(adata.obs[key].cat.categories, [to_rgb(i) for i in adata.uns[f"{key}_colors"]]))
+    col_dict = dict(zip(adata.obs[key].cat.categories, [to_rgb(i) for i in adata.uns[Key.uns.colors(key)]]))
 
     return np.array([col_dict[v] for v in adata.obs[key]])
 
@@ -42,7 +44,7 @@ def _position_cluster_labels(coords: np.ndarray, clusters: pd.Series, colors: np
     df = pd.DataFrame(coords)
     df["clusters"] = clusters.values
     df = df.groupby("clusters")[[0, 1]].apply(lambda g: list(np.median(g.values, axis=0)))
-    df = pd.DataFrame((r for r in df), index=df.index)
+    df = pd.DataFrame(list(df), index=df.index)
 
     kdtree = KDTree(coords)
     clusters = np.full(len(coords), fill_value="", dtype=object)

--- a/squidpy/pl/_utils.py
+++ b/squidpy/pl/_utils.py
@@ -202,14 +202,18 @@ def _points_inside_triangles(points: np.ndarray, triangles: np.ndarray) -> np.nd
 
 def _min_max_norm(vec: Union[spmatrix, np.ndarray]) -> np.ndarray:
     if issparse(vec):
-        vec = vec.A.squeeze()
+        if TYPE_CHECKING:
+            assert isinstance(vec, spmatrix)
+        vec = vec.toarray().squeeze()
     vec = np.asarray(vec, dtype=np.float64)
     if vec.ndim != 1:
         raise ValueError(f"Expected `1` dimension, found `{vec.ndim}`.")
 
     maxx, minn = np.nanmax(vec), np.nanmin(vec)
 
-    return np.ones_like(vec) if np.isclose(minn, maxx) else ((vec - minn) / (maxx - minn))
+    return (  # type: ignore[no-any-return]
+        np.ones_like(vec) if np.isclose(minn, maxx) else ((vec - minn) / (maxx - minn))
+    )
 
 
 def _ensure_dense_vector(fn: Callable[..., Vector_name_t]) -> Callable[..., Vector_name_t]:
@@ -236,6 +240,8 @@ def _ensure_dense_vector(fn: Callable[..., Vector_name_t]) -> Callable[..., Vect
                 raise TypeError(f"Unable to process `pandas.Series` of type `{infer_dtype(res)}`.")
             res = res.to_numpy()
         elif issparse(res):
+            if TYPE_CHECKING:
+                assert isinstance(res, spmatrix)
             res = res.toarray()
         elif not isinstance(res, (np.ndarray, Sequence)):
             raise TypeError(f"Unable to process result of type `{type(res).__name__}`.")
@@ -559,7 +565,7 @@ def _filter_kwargs(func: Callable[..., Any], kwargs: Mapping[str, Any]) -> Dict[
     return {k: v for k, v in kwargs.items() if k in style_args}
 
 
-def _dendrogram(data: np.array, method: str, **kwargs: Any) -> Tuple[List[int], List[int], List[int], List[int]]:
+def _dendrogram(data: np.ndarray, method: str, **kwargs: Any) -> Tuple[List[int], List[int], List[int], List[int]]:
     link_kwargs = _filter_kwargs(sch.linkage, kwargs)
     dendro_kwargs = _filter_kwargs(sch.dendrogram, kwargs)
 

